dash/include/dash/algorithm/Sort.h

```diff
@@ -0,0 +1,1177 @@
+  auto const nonstable_it = std::find(
+      p_borders.is_stable.cbegin(), p_borders.is_stable.cend(), false);
+
+  DASH_LOG_TRACE("psort__validate_partitions >");
+  // exit condition
+  return nonstable_it == p_borders.is_stable.cend();
+}
+
+template <typename LocalArrayT>
+inline void psort__calc_final_partition_dist(
+    std::vector<typename LocalArrayT::value_type> const& acc_partition_count,
+    LocalArrayT&                                         l_partition_dist)
+{
+  /* Calculate number of elements to receive for each partition:
+   * We first assume that we we receive exactly the number of elements which
+   * are less than P.
+   * The output are the end offsets for each partition
+   */
+  DASH_LOG_TRACE("< psort__calc_final_partition_dist");
+
+  auto const myid       = l_partition_dist.pattern().team().myid();
+  auto const nunits     = l_partition_dist.pattern().team().size();
+  auto       dist_begin = l_partition_dist.begin() + IND_DIST(nunits);
+  auto       supp_begin = l_partition_dist.begin() + IND_SUPP(nunits);
+
+  auto const n_my_elements =
+      std::accumulate(dist_begin, dist_begin + nunits, 0);
+
+  // Calculate the deficit
+  auto my_deficit = acc_partition_count[myid + 1] - n_my_elements;
+
+  dash::team_unit_t       unit(0);
+  dash::team_unit_t const last(nunits);
+
+  // If there is a deficit, look how much unit j can supply
+  for (; unit < last && my_deficit > 0; ++unit) {
+    auto const supply_unit = *(supp_begin + unit) - *(dist_begin + unit);
+
+    DASH_ASSERT_GE(supply_unit, 0, "invalid supply of target unit");
+    if (supply_unit <= my_deficit) {
+      *(dist_begin + unit) += supply_unit;
+      my_deficit -= supply_unit;
+    }
+    else {
+      *(dist_begin + unit) += my_deficit;
+      my_deficit = 0;
+    }
+  }
+
+  DASH_ASSERT_GE(my_deficit, 0, "Invalid local deficit");
+  DASH_LOG_TRACE("psort__calc_final_partition_dist >");
+}
+
+template <typename LocalArrayT, typename ElementType>
+inline void psort__calc_send_count(
+    PartitionBorder<ElementType> const&            p_borders,
+    std::vector<size_t> const&                     valid_partitions,
+    LocalArrayT&                                   partition_dist,
+    std::vector<typename LocalArrayT::value_type>& l_send_displs)
+{
+  using value_t = typename LocalArrayT::value_type;
+  DASH_LOG_TRACE("< psort__calc_send_count");
+
+  auto const           nunits = partition_dist.pattern().team().size();
+  std::vector<value_t> tmp_target_count(nunits + 1);
+
+  tmp_target_count[0] = 0;
+
+  auto l_target_count = &(partition_dist[IND_TARGET_COUNT(nunits)]);
+  auto l_send_count   = &(partition_dist[IND_SEND_COUNT(nunits)]);
+
+  auto const last_skipped = p_borders.is_skipped.cend();
+  auto       it_skipped =
+      std::find(p_borders.is_skipped.cbegin(), last_skipped, true);
+
+  auto it_valid = valid_partitions.cbegin();
+
+  std::size_t skipped_idx = 0;
+
+  while (std::find(it_skipped, last_skipped, true) != last_skipped) {
```

Comment:
Is skipping the first found `it_skipped` intentional? If so, please document

Reply:
We do not skip the first found `it_skipped` here. If there is a first `it_skipped` it is processed in this while loop.